File: app/src/main/java/com/intfocus/yhdev/scanner/StoreSelectorActivity.java
package com.intfocus.yhdev.scanner;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.SearchView;
import android.widget.TextView;

import com.intfocus.yhdev.R;
import com.intfocus.yhdev.base.BaseActivity;
import com.intfocus.yhdev.data.response.scanner.StoreItem;
import com.intfocus.yhdev.data.response.scanner.StoreListResult;
import com.intfocus.yhdev.net.ApiException;
import com.intfocus.yhdev.net.CodeHandledSubscriber;
import com.intfocus.yhdev.net.RetrofitUtil;
import com.intfocus.yhdev.util.FileUtil;
import com.intfocus.yhdev.util.K;
import com.intfocus.yhdev.util.URLs;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by lijunjie on 16/8/15.
 */
public class StoreSelectorActivity extends BaseActivity {
    private ListView mListView;
    private String cachedPath;
    private ArrayList<StoreItem> dataList = new ArrayList<>();
    private ArrayList<String> storeNameList = new ArrayList<>();
    private JSONObject cachedJSON;

    @TargetApi(Build.VERSION_CODES.KITKAT)
    @Override
    @SuppressLint("SetJavaScriptEnabled")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_store_selector);

        final LinearLayout mListHead = (LinearLayout) findViewById(R.id.store_list_head);
        TextView mSelectedItem = (TextView) findViewById(R.id.store_item_select);

        try {
            cachedPath = FileUtil.dirPath(mAppContext, K.kCachedDirName, K.kBarCodeResultFileName);
            cachedJSON = FileUtil.readConfigFile(cachedPath);
            JSONObject currentStore = cachedJSON.getJSONObject(URLs.kStore);

            mSelectedItem.setText(currentStore.getString("name")); // 已选项显示当前门店

            RetrofitUtil.getHttpService(getApplicationContext()).getStoreList(mUserSP.getString("user_num", "0"))
                    .compose(new RetrofitUtil.CommonOptions<StoreListResult>())
                    .subscribe(new CodeHandledSubscriber<StoreListResult>() {
                        @Override
                        public void onError(ApiException apiException) {

                        }

                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onBusinessNext(StoreListResult data) {
                            if (data.getData() != null) {
                                for (int i = 0, len = data.getData().size(); i < len; i++) {
                                    dataList.add(data.getData().get(i));
                                    storeNameList.add(data.getData().get(i).getName());
                                }
                            }

                            /*
                             * 搜索框初始化
                             */
                            SearchView mSearchView = (SearchView) findViewById(R.id.storeSearchView);
                            int searchEditId = mSearchView.getContext().getResources().getIdentifier("android:id/search_src_text", null, null);
                            TextView mSearchEdit = (TextView) findViewById(searchEditId);
                            mSearchEdit.setTextSize(14);
                            mSearchEdit.setPadding(0, 30, 0, 0);

                            /*
                             * ListView 初始化
                             */
                            mListView = (ListView) findViewById(R.id.listStores);
                            ListArrayAdapter mArrayAdapter = new ListArrayAdapter(mAppContext, R.layout.list_item_report_selector, storeNameList);
                            mListView.setAdapter(mArrayAdapter);
                            mListView.setTextFilterEnabled(true);

                            /*
                             * 搜索框事件监听
                             */
                            mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                                // 当点击搜索按钮时触发该方法
                                @Override
                                public boolean onQueryTextSubmit(String query) {
                                    return false;
                                }

                                // 当搜索内容改变时触发该方法
                                @Override
                                public boolean onQueryTextChange(String newText) {
                                    if (!TextUtils.isEmpty(newText)) {
                                        mListHead.setVisibility(View.GONE);
                                        mListView.setFilterText(newText);
                                    } else {
                                        mListHead.setVisibility(View.VISIBLE);
                                        mListView.clearTextFilter();
                                    }
                                    return true;
                                }
                            });

                            /**
                             *  用户点击项写入本地缓存文件
                             */
                            mListView.setOnItemClickListener(mItemClickListener);
                        }
                    });
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /*
     * listview 点击事件
     */
    private AdapterView.OnItemClickListener mItemClickListener = new AdapterView.OnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            try {
                TextView mSelector = (TextView) view.findViewById(R.id.reportSelectorItem);
                String selectedItem = mSelector.getText().toString();
                for (int i = 0; i < dataList.size(); i++) {
                    if (dataList.get(i).getName().equals(selectedItem)) {
                        JSONObject storeJson = new JSONObject();
                        storeJson.put("id", dataList.get(i).getId());
                        storeJson.put("name", dataList.get(i).getName());
                        cachedJSON.put(URLs.kStore, storeJson);
                        FileUtil.writeFile(cachedPath, cachedJSON.toString());
                    }
                }

                dismissActivity(null);
            } catch (Exception e) {
                e.printStackTrace();
            }
            dismissActivity(null);
        }
    };

    protected void onResume() {
        mMyApp.setCurrentActivity(this);
        super.onResume();
    }

    protected void onDestroy() {
        mWebView = null;
        super.onDestroy();
    }

    public class ListArrayAdapter extends ArrayAdapter<String> {
        private int resourceId;
        private List<String> items;

        public ListArrayAdapter(Context context, int textViewResourceId, List<String> items) {
            super(context, textViewResourceId, items);
            this.resourceId = textViewResourceId;
            this.items = items;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            String item = items.get(position).trim();
            LinearLayout listItem = new LinearLayout(getContext());
            String inflater = Context.LAYOUT_INFLATER_SERVICE;
            LayoutInflater vi = (LayoutInflater) getContext().getSystemService(inflater);
            vi.inflate(resourceId, listItem, true);
            TextView viewItem = (TextView) listItem.findViewById(R.id.reportSelectorItem);
            viewItem.setText(item);
            viewItem.setTextColor(getResources().getColor(R.color.black));
            viewItem.setBackgroundColor(Color.WHITE);

            return listItem;
        }
    }

    /*
     * 返回
     */
    public void dismissActivity(View v) {
        StoreSelectorActivity.this.onBackPressed();
        finish();
    }
}

File: app/src/main/java/com/intfocus/yhdev/scanner/StoreSelectorActivity.kt
package com.intfocus.yhdev.scanner

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.intfocus.yhdev.R
import com.intfocus.yhdev.base.BaseActivity
import com.intfocus.yhdev.data.response.scanner.StoreItem
import com.intfocus.yhdev.scanner.adapter.SelectorListAdapter
import com.intfocus.yhdev.scanner.presenter.StoreSelectorPresenter
import com.intfocus.yhdev.scanner.view.StoreSelectorView
import com.intfocus.yhdev.util.URLs
import kotlinx.android.synthetic.main.activity_store_selector.*

/**
 * ****************************************************
 * @author jameswong
 * created on: 17/10/19 下午5:43
 * e-mail: PassionateWsj@outlook.com
 * name:
 * desc:
 * ****************************************************
 */
class StoreSelectorActivity : BaseActivity(), StoreSelectorView {
    companion object {
        val RESULT_CODE_CHOOSE = 2
    }

    lateinit var presenter: StoreSelectorPresenter
    lateinit var adapter: SelectorListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_store_selector)

        initAdapter()
        initData()
        initListener()
    }

    private fun initAdapter() {
        store_item_select.text = intent.getStringExtra(URLs.kStore)
        adapter = SelectorListAdapter(this)
        listStores.adapter = adapter
    }


    private fun initData() {
        presenter = StoreSelectorPresenter(this, this)
        presenter.loadData()
    }

    private fun initListener() {
        // 隐藏软键盘
        ll_store_selector_container.setOnTouchListener { _, _ ->
            hideKeyboard()
            false
        }
        //　item 点击监听
        listStores.setOnItemClickListener { _, _, position, _ ->
            val intent = Intent()
            intent.putExtra(URLs.kStore, adapter.getSelectItem(position).name)
            intent.putExtra(URLs.kStoreIds, adapter.getSelectItem(position).id)
            setResult(RESULT_CODE_CHOOSE, intent)
            finish()
        }
        // 文本输入框内容监听
        storeSearchView.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(p0: Editable?) {
                presenter.loadData(p0.toString())
            }

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }
        })
        // 根据搜索内容 请求数据
        tv_store_selector_search.setOnClickListener {
            hideKeyboard()
            presenter.loadData(storeSearchView.text.toString())
        }
        // 上次选择内容 点击监听
        store_item_select.setOnClickListener { finish() }
    }

    /**
     * 获取数据成功
     */
    override fun onResultSuccess(data: MutableList<StoreItem>?) {
        adapter.setData(data)
    }

    /**
     * 获取数据失败
     */
    override fun onResultFailure(e: Throwable?) {

    }

    /**
     * 返回监听
     */
    fun backPress(v: View) {
        hideKeyboard()
        finish()
    }

    /**
     * 隐藏软件盘
     */
    fun hideKeyboard() {
        val imm = getSystemService(
                Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(this.currentFocus
                .windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
    }
}
